text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,284 @@
+ 
+The Ledger Enclave stores all attestation reports, as signed by the TEE vendor. Reports include information on what chaincode is running; what specific TEE is in use, including version number and hash of the chaincode, and including the public keys of the enclave. With the planned new key management scheme to support multiple endorsing peers, there may be many public keys instead of one. Upon creation each chaincode enclave generates public/private key pairs for signing and for encryption. The public signing key also denotes the chaincode enclave identity. In the new scheme, there will also be a public/private key pair and a symmetric state encryption key per FPC chaincode that are shared among all chaincode enclaves that run the same FPC chaincode using the key distribution protocol above.
+  
+ 
+## FPC Chaincode Lifecycle
+ 
+To illustrate how the FPC architecture works and how it ensures robust end-to-end trust, we describe the lifecycle of an FPC transaction. This assumes that all of the above described elements are already in place. Detailed explanations of that process are included below in the Reference-Level Explanation.
+ 
+* Step 1: Client invocation of the FPC Chaincode
+ 
+    The Client prepares the Invocation of an Enclave Chaincode by first encrypting the arguments of the Chaincode Invocation using the public key specific to a particular Chaincode. This encryption happens completely transparently using our FPC Client SDK extension. This Transaction Proposal is then sent to the Endorsing Peer where a corresponding Chaincode Enclave resides. Depending on the Endorsement Policy the client may perform this step with one or more Endorsing Peers and their respective Chaincode Enclaves. (For simplicity we will continue describing the process for a single Endorsing Peer.) The Peer forwards the Transaction Proposal to its FPC Chaincode running inside the Chaincode Enclave. Inside the Enclave, the FPC shim decrypts the Proposal and invokes the FPC Chaincode.
+ 
+* Step 2: Chaincode Execution
+ 
+    Having received and decrypted the Transaction Proposal, the FPC Chaincode processes the invocation according to the implemented chaincode logic. While executing, the chaincode can access the World State through getState and putState operations provided by the Chaincode Library shim. The Chaincode Library Shim fetches the state data from the peer and loads it into the chaincode enclave; then the chaincode library verifies that the received data is correct (e.g. is actual committed data) with the help of the Ledger Enclave. This protects the Chaincode from various attacks that might be possible in the case of a compromised Peer. (An example and explanation is included below.) We refer to our paper (see prior art section) for more details on this type of attacks.
```

Comment:
When the peer is compromised, it is possible that the attacker kills the peer and then restarts it. I would like to know what happens when the peer restarts. For example, I would like to know which enclaves are stateless/stateful. During the restart process, how each enclave works, etc...

Is it possible for the attacker to get some important data during the peer restart? May be not but have these been analyzed?

Reply:
A compromised host (peer) can indeed restart the peer (process) and all enclave. TEE's in general cannot prevent this. This is a good point, we should address this in the RFC. In principle, all enclaves are stateful, in particular, the trusted ledger enclave maintains the integrity metadata of the world state and every chaincode enclave maintains a unique identity which has been registered at our enclave registry.